text_collector_examples/storcli.py

```diff
@@ -12,81 +12,77 @@
 """
 
 from __future__ import print_function
+from datetime import datetime
 import argparse
+import collections
 import json
 import os
-import subprocess
 import shlex
-from dateutil.parser import parse
-import collections
-from enum import IntEnum
+import subprocess
 
 DESCRIPTION = """Parses StorCLI's JSON output and exposes MegaRAID health as
     Prometheus metrics."""
-VERSION = '0.0.2'
+VERSION = '0.0.3'
 
 storcli_path = ''
 metric_prefix = 'megaraid_'
 metric_list = {}
 metric_list = collections.defaultdict(list)
 
 
-class VD_State(IntEnum):
-    Optl = 0  # Optimal
-    Dgrd = 1  # Degraded
-    Pdgd = 2  # Partially Degraded
-    OfLn = 3  # Offline
-    Rec = 4  # Recovery
-    Cac = 5  # CacheCade
-
-
 def main(args):
     """ main """
     global storcli_path
     storcli_path = args.storcli_path
-    data = json.loads(get_storcli_json('/cALL show all J'))
-
-    # All the information is collected underneath the Controllers key
-    data = data['Controllers']
-
-    # try:
-    #     overview = status['Response Data']['System Overview']
-    # except KeyError:
-    #     pass
-
-    for controller in data:
-        response = controller['Response Data']
-        if response['Version']['Driver Name'] == 'megaraid_sas':
-            handle_megaraid_controller(response)
-        elif response['Version']['Driver Name'] == 'mpt3sas':
-            handle_sas_controller(response)
-
-    # print_dict_to_exporter({'controller_info': [1]}, controller_info_list)
-    # print_dict_to_exporter({'virtual_disk_info': [1]}, vd_info_list)
-    # print_dict_to_exporter({'physical_disk_info': [1]}, pd_info_list)
-    # print_all_metrics(vd_metric_list)
+    data = get_storcli_json('/cALL show all J')
+
+    try:
+        # All the information is collected underneath the Controllers key
+        data = data['Controllers']
+
+        for controller in data:
+            response = controller['Response Data']
+            if response['Version']['Driver Name'] == 'megaraid_sas':
+                handle_megaraid_controller(response)
+            elif response['Version']['Driver Name'] == 'mpt3sas':
+                handle_sas_controller(response)
+    except KeyError:
+        pass
+
     print_all_metrics(metric_list)
 
 
 def handle_sas_controller(response):
-    pass
+    (controller_index, baselabel) = get_basic_controller_info(response)
+    add_metric('healthy', baselabel, int(response['Status']['Controller Status'] == 'OK'))
+    add_metric('ports', baselabel, response['HwCfg']['Backend Port Count'])
+    try:
+        # The number of physical disks is half of the number of items in this dict
+        # Every disk is listed twice - once for basic info, again for detailed info
+        add_metric('physical_drives', baselabel,
+                   len(response['Physical Device Information'].keys()) / 2)
+    except AttributeError:
+        pass
+    add_metric('temperature', baselabel, int(response['HwCfg']['ROC temperature(Degree Celcius)']))
```

Comment:
codespell complaint:
```
./text_collector_examples/storcli.py:72: Celcius  ==> Celsius
```

Reply:
Yes - unfortunately this is in the name of the key of the JSON response of storcli... and it differs between storcli drivers... nothing I can change :/

See these two lines:
https://github.com/prometheus/node_exporter/blob/0255768e9e13c881169b24fc71c35203e3afe267/text_collector_examples/storcli.py#L72
https://github.com/prometheus/node_exporter/blob/0255768e9e13c881169b24fc71c35203e3afe267/text_collector_examples/storcli.py#L95

Reply:
That's terrible. :(

I suppose we could skip this file, or the whole examples dir in codespell config.

Reply:
I'd probably go with the hack, codespell doesn't have a very granular exception setup.

Reply:
Uhm that's ugly. There is `--ignore-words`. I hope they add 'directive' to ignore though, feature request/issue here: https://github.com/codespell-project/codespell/issues/66